Add 100/1000s buckets for prometheus workqueue histograms QueueLatencyKey and WorkDurationKey

youwalther65 opened the pull request:
This PR changes QueueLatencyKey and WorkDurationKey Prometheus histograms to contain 12 buckets instead of currently 10.
Controllers making many external requests in large clusters may have normal operating latency on the order of ~100s. Add buckets that cover the range, allowing metric backends to interpolate percentile estimates properly.
A good example is KCM HPA controller when talking to external metrics providers and using hundreds of HPA objects and external metrics..
This was already implemented in controller-runtime, see https://github.com/kubernetes-sigs/controller-runtime/issues/2625 and merged https://github.com/kubernetes-sigs/controller-runtime/pull/2638.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes [#126100](https://github.com/kubernetes/kubernetes/issues/126100)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
K8s components like KCM will now expose 100s and 1000s bucktes in histogram QueueLatencyKey and WorkDurationKey
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

youwalther65 committed:
Update metrics.go

Changed QueueLatencyKey and WorkDurationKey Prometheus histogram to contain 12 buckets

youwalther65 commented:
/sig instrumentation

dims commented:
/ok-to-test

dgrisonnet commented:
/assign
/triage accepted

dgrisonnet commented:
From an instrumentation standpoint, this change is reasonable, but I am wondering if it really makes sense to have bucket for like 10e-9, 10e-8, 10e-7. From my non-expert point of view, it seems way to low to really matter. So to compensate the addition of the two new buckets, maybe I would make sense to reduce the precision we have today?

@alculquicondor do you perhaps have the expertise on that particular topic?

youwalther65 commented:
> From an instrumentation standpoint, this change is reasonable, but I am wondering if it really makes sense to have bucket for like 10e-9, 10e-8, 10e-7. From my non-expert point of view, it seems way to low to really matter. So to compensate the addition of the two new buckets, maybe I would make sense to reduce the precision we have today?
> 
> @alculquicondor do you perhaps have the expertise on that particular topic?

That's a valid question. We should take all controller possibly using this code into account , but 10e-9 (1 nanosecond) is something I have never seen before. When we are talking about for example KCM work queue latencies etc. this is most probably in ms range (Millisecond, 10e-3).
Grafana dashboard will automatically adapt to the # of buckets anyway.

youwalther65 commented:
@alculquicondor @dgrisonnet An alternative approach to base 10 logarithmic would be using kube-apiserver endpoints approach for example `request_duration_seconds` [here](https://github.com/pawbana/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L91-L104) with:
````
Buckets: []float64{0.005, 0.025, 0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3,
				4, 5, 6, 8, 10, 15, 20, 30, 45, 60},
```

youwalther65 commented:
@serathius @mengjiao-liu What's your opinion on that? Thx in advance!

serathius commented:
> Controllers making many external requests in large clusters may have normal operating latency on the order of ~100s.

What's the SLO we provide for controller operation latency? https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md

youwalther65 commented:
> > Controllers making many external requests in large clusters may have normal operating latency on the order of ~100s.
> 
> What's the SLO we provide for controller operation latency? https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md

As far as I understand K8s SLO there is not yet an SLO for controller operations. The quote I made is directly copied over from controller-runtime PR [here](https://github.com/kubernetes-sigs/controller-runtime/pull/2638)

serathius commented:
I bring the SLO because it should be a clearly agreed-upon contract about the boundary of what latency is acceptable, after which we should treat higher latency as just an error. Adding more buckets to a metric doesn't come without a cost; if it didn't, we wouldn't need to use histograms at all. We need to think about sensible latency ranges that administrators care about.

As brought up by @dgrisonnet, do we really care about reconciliation taking below microseconds? I don't think that's even possible, as it usually requires communication to the API server, which adds milliseconds of network delay. Do we really care about a controller taking 100s to reconcile an object? Maybe, assuming that the expected reconciliation time was 10-100s. However, if the average reconciliation takes 100ms, why would anyone care if something took 10, 20, 50, or 100s?

To know what is an acceptable boundary for monitoring latency, you need to know what is the acceptable range of values that administrators/automation should care about. Please don't just add more buckets to a metric, as you just might be increasing the cost of monitoring by 20% without any consideration.

youwalther65 commented:
I completely agree in everything you wrote @serathius.
The only data points I have is a) the controller-runtime change, hoping the values they used are reasonable and b) I see some customer data where HPA KCM work queue is constantly hitting the 10s line, so I only know the latency is 10s or higher.

youwalther65 commented:
The [workqueue](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/component-base/metrics/prometheus/workqueue/metrics.go) package I ' propose to change is used 3 times - for [kube-apiserver](https://github.com/kubernetes/kubernetes/blob/release-1.31/cmd/kube-apiserver/app/server.go#L47) as well as [kube-controller-manager](https://github.com/kubernetes/kubernetes/blob/release-1.31/staging/src/k8s.io/controller-manager/app/serve.go#L34) and cloud-controller manager.
So these are the components affected by the additional buckets for upper histogram end 100s and 1000s. On the lower end I haven't seen any occurance of 1e-08, 1e-07 and 1e-06. So I would even go one step further and propsoe to remove  at least 1e-08 , 1e-07 and add 100s, 1000s so we have the same number of buckets and same metrics cost.

youwalther65 committed:
Update metrics.go

reduced bucket for workqueue latency and workDuration count back to 10, but removed 10e-9 and 10e-8 bucket

alculquicondor commented:
Not sure why I was tagged here. This is a question for instrumentation.

But, I agree that `1e-9` doesn't sound like a useful bucket.

RiskyAdventure commented:
@serathius @youwalther65 
Thank you for your help. I was interested in the range values for the workqueue queue duration and the work duration metrics. I used the count/sum (avg) vs. the bucketed metrics.

`rate(workqueue_queue_duration_seconds_sum{job="$job"}[5m]) / rate(workqueue_queue_duration_seconds_count{job="$job"}[5m]))`

These values were in the range of minutes for things like the GC controller, when running large batch jobs as an example.
 
![Screenshot 2024-10-30 at 10 30 26 AM](https://github.com/user-attachments/assets/d8cad949-0f28-4a76-aca5-9713ab1c67e4)

My use case is I'm trying to identify user agents that are slow to process. Having metrics in the sub-seconds scale I don't find overly useful for the following two reasons. 

- Having only two values above one second (1,10) would skew the distribution of values under 10 seconds, making it possibly misleading for the data points I'm looking for. (If I'm thinking about this correctly)

- Having the range so limited makes it difficult to tell if the key in the workqueue is really impacted. As a user story let say I have 2 controllers at 11 seconds of work duration, and one at one 60 seconds, I'm not able to understand this dynamic unless I go to the logs and check all three. There could be a lot of value in shifting this metric in the troubleshooting process.

With values in the upper range being in the minutes, I'd like to make the case it would be useful to start the bucket values at one second, keeping the same number of buckets. At least in my use case there is diminishing returns for more buckets, as I just want to know which controller might be having a processing issue. The caveat being, 100ms might be a better bottom floor if such times are needed for regression testing.

youwalther65 commented:
/remove-lifecycle stale

youwalther65 commented:
There is still a need to implement changes here. Just recently saw other customers limited by the 10s bucket.

I have two ideas to proceed here:

1. Attend a SIG Instrumentation meeting to discuss how to proceed
2. Get in direct contact with SIG Instrumentation maintainers at KubeCon Europe in April

youwalther65 commented:
Still in discussion with @dgrisonnet if we will wait for native histogram support or push this PR.

kmala commented:
In large clusters, some controllers(ex: daemonset, endpoint_slice) would take more time where list operations are made during the reconciliation. The current buckets and also the multiples won't help to know the performance of the controller because we can't differentiate if there is a performance degradation when the controllers start taking 7 seconds to reconcile instead of 2 seconds. I think we need to also think about the values of the buckets as jump of 1 sec to 10 seconds is pretty high.

youwalther65 commented:
/remove-lifecycle stale

youwalther65 commented:
> Note, this will be addressed by [kubernetes/enhancements#5649](https://github.com/kubernetes/enhancements/pull/5649)

@serathius This sounds interesting. But I somehow feel, it might not help in all the described use cases. Sometimes there are external dependencies like 3rd party metric provider which slow down processing of requests inisde controllers like HPA controller. How does this approach could help here?
@dgrisonnet proposed [Native Histograms](https://github.com/prometheus/docs/blob/main/docs/specs/native_histograms.md) (still EXPERIMENTAL in Prometheus as of today) to fix the visibility aspect of my issue.
If I am missing someone within the context of KEP #5649 , please let me know, thx.

youwalther65 committed:
Merge branch 'kubernetes:master' into master
